Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift

```diff
@@ -121,12 +121,59 @@ fileprivate extension TrialOrIntroPriceEligibilityChecker {
+    func getIntroEligibility(with receiptData: Data,
+                             productIdentifiers: [String],
+                             completion: @escaping ReceiveIntroEligibilityBlock) {
+        if #available(iOS 11.2, macOS 10.13.2, macCatalyst 13.0, tvOS 11.2, watchOS 6.2, *) {
+            self.productsWithIntroOffers(productIdentifiers: productIdentifiers) {
+                self.getIntroEligibility(with: receiptData,
+                                         productIdentifiers: productIdentifiers,
+                                         productIdsToIntroEligibleStatusFromApple: $0,
+                                         completion: completion)
+            }
+        } else {
+            self.getIntroEligibility(with: receiptData,
+                                     productIdentifiers: productIdentifiers,
+                                     productIdsToIntroEligibleStatusFromApple: [:],
+                                     completion: completion)
+        }
+    }
+
+    @available(iOS 11.2, macOS 10.13.2, macCatalyst 13.0, tvOS 11.2, watchOS 6.2, *)
+    func productsWithIntroOffers(productIdentifiers: [String], completion: @escaping ReceiveIntroEligibilityBlock) {
+        self.productsManager.products(withIdentifiers: Set(productIdentifiers)) { products in
+            let eligibility: [(String, IntroEligibility)] = Array(products).compactMap {
+                guard $0.introductoryPrice != nil else {
+                    return nil
+                }
+                return ($0.productIdentifier, IntroEligibility(eligibilityStatus: .eligible))
+            }
+            var productIdsToIntroEligibleStatus: [String: IntroEligibility] = [:]
+            eligibility.forEach { (productId, eligibility) in
+                productIdsToIntroEligibleStatus[productId] = eligibility
+            }
+            completion(productIdsToIntroEligibleStatus)
+        }
+    }
+
+}
+
+private extension TrialOrIntroPriceEligibilityChecker {
+
+    func getIntroEligibility(with receiptData: Data,
+                             productIdentifiers: [String],
+                             productIdsToIntroEligibleStatusFromApple: [String: IntroEligibility],
+                             completion: @escaping ReceiveIntroEligibilityBlock) {
+        // Remove any productIds we already have intro pricing for so we don't try to fetch them from the backend.
+        let idsToFetchFromBackend = productIdentifiers.filter { productIdsToIntroEligibleStatusFromApple[$0] == nil }
+        if idsToFetchFromBackend.isEmpty {
+            completion(productIdsToIntroEligibleStatusFromApple)
+            return
+        }
+
+        self.backend.fetchIntroEligibility(appUserID: self.appUserID,
+                                           receiptData: receiptData,
+                                           productIdentifiers: productIdentifiers) { backendResult, maybeError in
```

Comment:
if `productIdentifiers` contains items that have `introductoryPrice == nil`, we're still sending them to the backend here, right? 
Ideally we only send the ones that have intro price to the backend to save on compute time

Reply:
Just to be sure I'm on the same page:

Assume we have `productIdentifiers` which is a list of 4 ids:
`productIdentifiers = [A has introductoryPrice, B does not, C does not, D has introductoryPrice]`
1.  We fetch on device and notice `A` and `D` have `introductoryPrice`
2. `productIdsToIntroEligibleStatusFromApple` now contains `["A" : A, "D": D]`
3.  We pass `productIdentifiers` and `productIdsToIntroEligibleStatusFromApple` to `getIntroEligibility`
4. We remove any products that have `introductoryPrice` from `productIdentifiers` in line 168, so now we have:
`idsToFetchFromBackend == [B, C]` and
`productIdsToIntroEligibleStatusFromApple` == ["A" : A, "D": D]`
We pass `[B, C]` to the backend to check on.

Looks like I forgot to use `idsToFetchFromBackend` for the call to `self.backend.fetchIntroEligibility`